src/include/daos_srv/control.h

```diff
@@ -14,21 +14,109 @@
 #include <stdint.h>
 #include <stdbool.h>
 #include <assert.h>
+#include <string.h>
 
-/*
+/**
  * Space separated string of CLI options to pass to DPDK when started during
  * spdk_env_init(). These options will override the DPDK defaults.
  */
 extern const char *
 dpdk_cli_override_opts;
 
-/* Device state flags */
+/** Device state flags */
 #define NVME_DEV_FL_PLUGGED	(1 << 0)
 #define NVME_DEV_FL_INUSE	(1 << 1) /* Used by DAOS (present in SMD) */
 #define NVME_DEV_FL_FAULTY	(1 << 2)
 #define NVME_DEV_FL_IDENTIFY	(1 << 3) /* SSD being identified by LED activity */
 
-/*
+/** Device state combinations */
+#define NVME_DEV_STATE_NORMAL (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_INUSE)
+#define NVME_DEV_STATE_FAULTY (NVME_DEV_STATE_NORMAL | NVME_DEV_FL_FAULTY)
+#define NVME_DEV_STATE_NEW (NVME_DEV_FL_PLUGGED)
+
+static inline char *
+nvme_state2str(int dev_state)
+{
+	switch (dev_state) {
+	case (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_INUSE | NVME_DEV_FL_FAULTY | NVME_DEV_FL_IDENTIFY):
+		return "EVICTED|IDENTIFY";
+	case (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_INUSE | NVME_DEV_FL_FAULTY):
+		return "EVICTED";
+	case (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_INUSE | NVME_DEV_FL_IDENTIFY):
+		return "IDENTIFY";
+	case (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_INUSE):
+		return "NORMAL";
+	case (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_FAULTY | NVME_DEV_FL_IDENTIFY):
+		return "NEW|EVICTED|IDENTIFY";
+	case (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_FAULTY):
+		return "NEW|EVICTED";
+	case (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_IDENTIFY):
+		return "NEW|IDENTIFY";
+	case NVME_DEV_FL_PLUGGED:
+		return "NEW";
+	case (NVME_DEV_FL_INUSE | NVME_DEV_FL_FAULTY | NVME_DEV_FL_IDENTIFY):
+		return "UNPLUGGED|EVICTED|IDENTIFY";
+	case (NVME_DEV_FL_INUSE | NVME_DEV_FL_FAULTY):
+		return "UNPLUGGED|EVICTED";
+	case (NVME_DEV_FL_INUSE | NVME_DEV_FL_IDENTIFY):
+		return "UNPLUGGED|IDENTIFY";
+	case NVME_DEV_FL_INUSE:
+		return "UNPLUGGED";
+	case (NVME_DEV_FL_FAULTY | NVME_DEV_FL_IDENTIFY):
+		return "UNPLUGGED|NEW|EVICTED|IDENTIFY";
+	case NVME_DEV_FL_FAULTY:
+		return "UNPLUGGED|NEW|EVICTED";
+	case NVME_DEV_FL_IDENTIFY:
+		return "UNPLUGGED|NEW|IDENTIFY";
+	case 0:
+		return "UNPLUGGED|NEW";
+	}
+
+	return "UNKNOWN";
+}
+
+static inline int
+nvme_str2state(char *state_str)
+{
+	if (strcmp(state_str, "EVICTED|IDENTIFY") == 0) {
+		return (NVME_DEV_FL_PLUGGED | NVME_DEV_FL_INUSE | NVME_DEV_FL_FAULTY |
+			NVME_DEV_FL_IDENTIFY);
```

Comment:
Same as above. We can decipher the bit values by parsing a copy of the string with strtok, and maybe avoid hard-coding a table like this.

Reply:
as above